### PR TITLE
BENCH: update/rewrite lobpcg benchmark 

### PR DESCRIPTION
**Pull request by lobpcg**
Update/rewrite sparse_linalg_lobpcg.py to make it more reasonable and recent.

#### Reference issue
Closes [gh-18568] and partially [gh-18939]. Closes [gh-18940] if the decision would be made to do it in this PR as a free bonus.

#### What does this implement/fix?
Addressing  [gh-18568], change the examples to make more sense, e.g., use the tests that actually converge, and checking the convergence. The current benchmark runs lobpcg and the dense eigensolver eigh on two generalized eigenproblems with analytically known eigenvalues, called Sakurai and Mikota_pair. 

Major enhancements for all cases: 
- Introducing assert of accuracy of computed eigenvalues in a uniform way so that a benchmark test fails in case of poor/no convergence. 
- Adding the only other available sparse eigensolver to the benchmark - the ARPACK-based eigsh.
- Tuning matrix sizes, maxit, and tol for all the tests to make sense. 
- Introducing matrix-free calls of all the matrices involved using functions that perform the dot-product of the matrix by an input matrix without constructing the matrix explicitly.

Test specific enhancements: 
- In Sakurai, in the sparse matrix pair A and B, the matrix A is the identity, and the matrix B gets ill-conditioned with its size n growing, leading to divergence in lobpcg for one of the tested cases n=400. The lobpcg algorithm is much more sensitive to ill-conditioning in B compared to that in A, so to simplify the convergence issues, we drop the generalized problem and solve instead a regular eigenproblem for a single matrix, B. Since the matrix A is the identity, the original eigenvalues get inverted. The regular eigenproblem allows easy adding the ARPACK-based eigsh to the test. Convergence/accuracy of lobpcg and eigsh has been analyzed. It turns out that eigsh does not converge for n>120 no matter the tolerance, so the tests now runs for n=50 and 100 and the tolerance and the max number of iterations are chosen to ensure a reasonably fair comparison of the performance of lobpcg vs. eigsh. 
- A new test with Sakurai is added that runs  lobpcg and eigsh on the matrix inverse taking into account that the Sakurai matrix is banded - using solvers for banded matrices and for a fair comparison calling a dense eigensolver for banded matrices rather than the generic dense eigh. 
- In Mikota_pair, in the matrix pair A and B, the matrix B is diagonal, and both matrices A and B get ill-conditioned with its size n growing. In the original tests, the matrices are constructed as dense, although A is tridiagonal, and B is diagonal. In the modified tests, eigsh runs on the matrix A inverse taking advantage of the fact that it is tridiagonal - using solvers for banded matrices. Inversion of A is chosen vs. a cheaper inversion of the diagonal matrix B because convergence of ARPACK in eigsh on inv(A) B in this example is much faster than that on inv(B) A. The lobpcg call goes a step further and uses  inv(A) in float32 precision for preconditioning, while ARPACK does not allow direct preconditioning. 

Partially addressing [gh-18939], Sakurai and Mikota matrices:
1. are being moved to a public API 
2. more matrix formats introduced in their definitions
3. these new formats are being use in benchmarking

Addressing [gh-18940], the Laplacian matrices has also been written in the style similar to that of Sakurai and Mikota matrices.  

I would be happy to remove this yet currently unfinished new Laplacian example as well as move the 2 modified Sakurai and Mikota matrices examples back to benchmarking from the public linalg temporarily postponing these changes to subsequent PRs, if the reviews of this PR could be made timely so that I could make all these in subsequent PR within a reasonable time frame.

### CODE REVIEW

**Comment by lobpcg** [comment]
Does anyone know why circle-ci bench no longer runs?

File: benchmarks/benchmarks/sparse_linalg_lobpcg.py
Comment (h-vetinari) [inline benchmarks/benchmarks/sparse_linalg_lobpcg.py:16]: Can you wrap the docstring correctly please?
```suggestion
    """
    Example turns a generalized eigenproblem for the matrix pair A and B
    T. Sakurai, H. Tadano, Y. Inadomi and U. Nagashima
    A moment-based method for large-scale generalized eigenvalue problems
```

**Review by h-vetinari — COMMENTED** [review]
AFAIU the benchmarks are not run per PR, but on a cron-job. CC @tupui

**Comment by tupui** [comment]
Benchmarks are run with CircleCI and should always run. I am not sure why in this case CircleCI did not run though 🤷‍♂️ 

The relevant config starts here:
https://github.com/scipy/scipy/blob/4bf18d332455f498b760e62657d8203d61b3c496/.circleci/config.yml#L135

@lobpcg please can you name properly your commit (and edit the title of your PR too)? For anyone having a look, it's difficult to see what is done. I remember you said you use VSCode, you should be able to change the "title" of the commit. I suppose you are only updating the message part.

Screenshot 2023-07-10 at 12 45 49

**Comment by h-vetinari** [comment]
> For anyone having a look, it's difficult to see what is done.

Yeah, I had the same issue in #17675 until I saw that the description is below the main commit message (_that_ is what should be the main commit message; the "update <file>" is irrelevant, or actively detrimental).

**Comment by h-vetinari** [comment]
> Benchmarks are run with CircleCI and should always run. I am not sure why in this case CircleCI did not run though 🤷‍♂️

Thanks for chiming in @tupui. If this keeps happening I guess we'll have to open an issue to investigate more thoroughly...

**Comment by lobpcg** [comment]
> > Benchmarks are run with CircleCI and should always run. I am not sure why in this case CircleCI did not run though 🤷‍♂️
> 
> Thanks for chiming in @tupui. If this keeps happening I guess we'll have to open an issue to investigate more thoroughly...

@h-vetinari @tupui  CircleCI is surely a must in this PR that only edits a benchmark...

**Comment by tupui** [comment]
Maybe it's because of the draft mode?? But I thought I recalled seeing draft PRs running on CircleCI. Maybe something changed there.

**Comment by lobpcg** [comment]
> Maybe it's because of the draft mode?? But I thought I recalled seeing draft PRs running on CircleCI. Maybe something changed there.

I only turned the draft mode today. It was not working prior to that too.

**Comment by h-vetinari** [comment]
It seems CircleCI as a whole isn't running in this PR...? 🤔

**Comment by lobpcg** [comment]
Please help to make cicrle-ci running! 

I have logged in to
https://app.circleci.com/pipelines/github/lobpcg/scipy 
but do not know how to use it - maybe I need to configure it somehow? Or is it independent from the scipy repo one?

**Comment by tupui** [comment]
There is nothing to do on your side.

I don't know what is happening here to be honest. What you could try is to make another PR... it's really never happening.

**Comment by lobpcg** [comment]
No circle-ci in the new PR too, https://github.com/scipy/scipy/pull/18874 :-(

**Comment by tupui** [comment]
I have no clue sorry... I am not seeing everything on CircleCI so there is not much more I can do. @rgommers would you know why CircleCI is not being run there?

**Comment by rgommers** [comment]
I can't find anything in CI config, repo settings or CI logs that seems related. I pushed a commit merging with main to see if that would trigger, but it doesn't. So I really have no idea. @lobpcg it shouldn't block merging this PR when it's ready - I am happy to verify it locally.

**Comment by lobpcg** [comment]
Since I only modify the benchmark (and move a few functions from it to _special_matrices.py) the only helpful ci is circle-ci, the one running the benchmarks and the one producing the documentation. I cannot run them locally without a very major effort.

At some point during my previous PR editing someone suggested that I configure and run cirle-ci myself rather than using the scipy quota, so I logged in to https://app.circleci.com/pipelines/github/lobpcg/scipy not really knowing what I am doing. Maybe this prevents the default scipy circle-ci running? 

It appears that I may be able to start circle-ci manually from https://app.circleci.com/pipelines/github/lobpcg/scipy ...

**Comment by tupui** [comment]
Ah yes, that could be it. I've read that this is an issue CircleCI is picking up your fork instead. So yes maybe removing all that would help.

You can also run the benchmarks locally and post results here.

**Comment by lobpcg** [comment]
While waiting for reviews I am trying to add a new special matrix - the Laplacain, to be used in benchmarking and examples

**Comment by j-bowhay** [comment]
One thing to consider is that adding to the public API (ie `linalg`) will probably make it much harder to find a reviewer as things then have to be reviewed far more rigorously. For the sake of not holding up the benchmark improvements, you might be best putting the code you have written for generating special matrices in the benchmark file. Once this is merged you can raise an issue explaining why these should have a spot in the public api.

**Comment by h-vetinari** [comment]
> While waiting for reviews I am trying to add a new special matrix - the Laplacain, to be used in benchmarking and examples

Since this PR is seeing regular commits, it was not obvious to me that at all that it's ready for review.

Where are you considering to add this Laplacian? If it's benchmark-only then it might still work for this PR, but otherwise it sounds like we should keep those things separate

**Comment by lobpcg** [comment]
@j-bowhay @h-vetinari I have opened the PR for review from @larsoner and @ilayn 3 days ago and have only made one (although it took a few commits) stylistic improvement to the 2 modified examples. This PR is waiting for review. 

My previous similar PR has taken several months to get reviews, so the normal practice of incremental PRs does not quite work in my case and instead I try to put more beef in every single PR. I would be happy to remove the yet unfinished new example - the Laplacian - and move the 2 modified examples back to benchmarking from the public linalg temporarily postponing these changes to subsequent PRs, if the reviews of this PR could be made timely so that I could make all these subsequent PR within a reasonable time frame. But I cannot afford waiting months for every PR in a sequence.  So many bugs so little time.

**Comment by lobpcg** [comment]
> @lobpcg as we discussed before in those PRs, I think the most optimal way is to somehow get your system to be able to compile and test the code on your local system. Otherwise we have to rely on Github CI/CD infrastructure which is making things quite difficult for all parties because like yourself we also have quite limited time and it becomes very difficult to watch you modify files manually probably here. Then you can catch the mistakes much easier and at your own convenience by running the tiny percent of the test suite without waiting 30 mins each time for the possible failures.
> 
> Please let me know if you need any support or guidance and we would be happy to help you with it here or offline. That would cut down the review time significantly.

@ilayn thank you for your kind offer to help! Yes, I run MS Win 11 with MS libraries and compilers, typically unsupported by open source. My last attempt has been to install the benchmark locally only to find out that it requires configuring an interface to BLAS/LAPACK, which I do have locally, but unable to find any documentation how to configure in benchmark, so give up. I have no access to any Linux, and installing/maintaining cygwin or a separate Linux instance may not worth the time only for my occasional PRs. 

Unable to run benchmarking locally, it makes little sense to keep a local fork either, so in this PR, I edit files directly using GitHub web interface. It makes commits very messy, so reviewers cannot see well the history and need to review only the current snapshot, indeed. I try minimizing other inconveniences to scipy maintainers, using [skip azp] [skip actions] [skip cirrus] on most of my commits and currently running my own https://app.circleci.com/pipelines/github/lobpcg/scipy?branch=patch-1 (which I thought was free credits but now I see they want to charge me $ 165.00 for this month...).  

The bottom line is that I am unsure how to improve my setup without wasting too much time.

**Comment by lobpcg** [comment]
I have updated the PR description to more fully describe the changes. Could someone please review this PR?

File: scipy/linalg/_special_matrices.py
Comment (h-vetinari) [inline scipy/linalg/_special_matrices.py:1403]: Why is this a class? All other matrices are constructed by a direct function call. The `__init__` + `callable` combination does the same in principle.

It should IMO not be a class, and if it has to be a class, this would need much more discussion (aside from a capitalized name).

**Review by h-vetinari — CHANGES_REQUESTED** [review]
I think mixing the benchmark clean up with new matrices is not a good idea. The benchmark fixes would probably mergeable much more quickly.

Using a class-based design would be a first in the special matrices module, and would need a very compelling reason for using a very different API.

File: scipy/linalg/_special_matrices.py
Comment (lobpcg) [reply scipy/linalg/_special_matrices.py:1403]: I just wanted to have a good looking way to call various objects, like mikota.callable, which the class natively provides. I don't know python well enough to figure out other similar good looking alternatives and of course I am open to constructive specific suggestions.

**Comment by lobpcg** [comment]
> I think mixing the benchmark clean up with new matrices is not a good idea. The benchmark fixes would probably mergeable much more quickly.
> 
> Using a class-based design would be a first in the special matrices module, and would need a very compelling reason for using a very different API.

@h-vetinari As I have already written above, I am OK with splitting this PR into 2 parts: benchmark fixes with matrices defined privately in benchmarks and moving the matrix definitions from benchmark to public special matrices, if necessary. And adding the Laplacian matrix (already committed) and benchmark (finished but not yet committed) on the top of that as a 3d PR. Only takes more of my time and drags this for (much?) longer.

**Comment by h-vetinari** [comment]
> Only takes more of my time and drags this for (much?) longer since all these 3 hypothetical PRs need to be done sequentially one after the other as far as I can see.

That's not true. The more straight-forward the PR, the faster it can get merged. Many unrelated changes in a single PR slow down review, not speed it up (even moreso if you're introducing fundamentally different APIs).

Also, most of the people here work on scipy purely as volunteers; while we try to be quick, it can happen that things take a while.

File: scipy/linalg/_special_matrices.py
Comment (h-vetinari) [reply scipy/linalg/_special_matrices.py:1403]: > good looking way

That's extremely subjective. So far, all matrices here can be simply generated with `X = name_of_the_matrix(...)`. I can see canonical matrix pairs as being constructed through `X, Y = name_of_the_pair(...)`, but I don't understand why you want to have something like `mikota.callable`. What's wrong with `mikota(...)`?

> and of course I am open to constructive specific suggestions.

It's you who wants/needs to add these functions; I don't understand your needs as well as you do, and I'm not so familiar with the linalg code overall (but at least I'm trying to give your PRs _some_ review...); that said, my specific suggestion is: barring extremely compelling evidence to the contrary (i.e. not just "good looking"), follow the patterns already existing in that module (resp. how those matrices are used in scipy).

**Comment by steppi** [comment]
> @h-vetinari OK. Should I wait for another reviewer to concur with your request to split this PR into 3 parts, if I interpret your request correctly, or just make the change? Or maybe I should preferably start a new clean PR with phase 1 - just benchmarking fixes with only the benchmarking module involved? Turning this one into "in progress" so far?

I concur with @h-vetinari's request to split this into 3 PRs. There's precedent for large PRs to remain open for years due to lack of reviewer bandwidth. In my experience, time to review can scale non-linearly with the size of the PR.

Also, @lobpcg, if you're having trouble getting a dev environment set up, have you considered using a [bootable usb linux install](https://www.groovypost.com/howto/best-usb-bootable-linux-distros/)? Someone could probably set one up for you that already has a SciPy dev environment configured. You would just have to install it to a usb drive and could boot from that without having to install Linux on one of your machines.

File: scipy/linalg/_special_matrices.py
Comment (lobpcg) [reply scipy/linalg/_special_matrices.py:1403]: @h-vetinari Most if not all iterative solvers in sparse scipy are matrix-free, i.e., do not actually need the matrix stored explicitly in memory, while this is what all the present functions produce in `X = name_of_the_matrix(...)`. E.g., `mikota.callable` is a callable object rather than the Mikota matrix currently produced by `mikota(...)`

Comment (h-vetinari) [reply scipy/linalg/_special_matrices.py:1403]: If you want to introduce a fundamentally new mechanism, a PR is not the place to discuss this. Please open an issue with the proposed design (including what problems it solves, what is necessary to solve them, why the design was chosen over possible alternatives, etc.), then we can discuss if/how we want to support this, and how the API should look. 

It doesn't have to be a specification down the the last detail, but it should at least present the high level requirements and considerations. Your branch can serve as an implementation reference and the benchmark numbers would be supporting evidence that this is worth the maintenance cost.

**Comment by lobpcg** [comment]
All activity gets transferred to https://github.com/scipy/scipy/pull/18954. I'll keep this one as a draft until everything is done.

**Comment by lobpcg** [comment]
> Also, @lobpcg, if you're having trouble getting a dev environment set up, have you considered using a [bootable usb linux install](https://www.groovypost.com/howto/best-usb-bootable-linux-distros/)? Someone could probably set one up for you that already has a SciPy dev environment configured. You would just have to install it to a usb drive and could boot from that without having to install Linux on one of your machines.

@steppi @ilayn yes, I think that it could be useful not only for me but in general to have a public bootable image with everything needed pre-installed to be booted into locally or/and on one of popular cloud providers, may be a quick tune if needed of one of already available free public images with dev tools.

**Comment by j-bowhay** [comment]
@lobpcg Another possibility is to use github code spaces. We have instructions for that here https://scipy.github.io/devdocs/dev/dev_quickstart.html#other-workflows . The downside being your account only has limited number of free hours

**Pull request closed without merging**
